### PR TITLE
Etcd memory and cpu requests

### PR DESCRIPTION
**Pull request by integrii**
Some of our etcd pods end up with etcd memory use that far exceeds the request defaults kops applies.  This change adds the ability to specify the pod memory and CPU requests on a per-etcd-cluster basis so that we can up our requests.  It also adds a default 200m CPU to etcd clusters named `main`, and 100m otherwise.  100Mi of memory is always assigned by default.  This should match prior functionality.

Two new options open up for etcd cluster configuration.  `cpuRequest` and `memoryRequest`.  These, when set, of course override the defaults described here.

There are plenty of things I can see that could be improved, like using a constant for the default memory and CPU.  I just wasn't sure where to put it.  It took me about two days to find all these parts of the code, work around bugs with things like api machinery.  Feel free to let me know if something needs refactored or adjusted.

Thanks!

### CODE REVIEW

**Comment by integrii** [comment]
@ihoegen 😻

**Comment by ihoegen** [comment]
I think you're going to need to add this to 
```
k8s.io/kops/pkg/apis/kops/v1alpha1/cluster.go
k8s.io/kops/pkg/apis/kops/v1alpha2/cluster.go
```

**Comment by integrii** [comment]
I added the config lines to the other cluster config versions, then I hit #6314 when building api machinery.  I unset my `GOBIN` to bypass the error for now.

Once that worked, I get this error:

```
E0108 22:15:49.365750   43694 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kops/pkg/apis/kops.EtcdClusterSpec -> k8s.io/kops/pkg/apis/kops/v1alpha1.EtcdClusterSpec
E0108 22:15:49.365777   43694 conversion.go:668]   the following fields need manual conversion:
E0108 22:15:49.365782   43694 conversion.go:670]       - MemoryRequest
E0108 22:15:49.365786   43694 conversion.go:670]       - CPURequest
/Users/egreer200/go/bin/conversion-gen  --skip-unsafe=true --input-dirs k8s.io/kops/pkg/apis/kops/v1alpha2 --v=0  --output-file-base=zz_generated.conversion \
		 --go-header-file "hack/boilerplate/boilerplate.go.txt"
E0108 22:15:51.688471   43695 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kops/pkg/apis/kops.EtcdClusterSpec -> k8s.io/kops/pkg/apis/kops/v1alpha2.EtcdClusterSpec
E0108 22:15:51.688496   43695 conversion.go:668]   the following fields need manual conversion:
E0108 22:15:51.688500   43695 conversion.go:670]       - MemoryRequest
E0108 22:15:51.688504   43695 conversion.go:670]       - CPURequest
```

How do I add said manual conversion?

**Comment by integrii** [comment]
I added the stuff genereated from the apimachinery run - was that right??

**Comment by integrii** [comment]
lol - did the files that it generates fail the `go fmt` test?

**Comment by ihoegen** [comment]
Yeah, looks like it's because of some whitespace. Try running `make gofmt` first

**Comment by integrii** [comment]
i just ran `go fmt` on everything and found nothing.  `make gofmt` does not find any changes, either. I will take a closer look at the errors here tomorrow.

image

**Comment by integrii** [comment]
@ihoegen I gave you access to my branch.  Can you run a format on it pretty please?  I'm guessing there is a tooling difference that will be annoying to find...

**Comment by ihoegen** [comment]
Gotta fix the Travis CI tests now: https://travis-ci.org/kubernetes/kops/jobs/477460494

Bunch of stuff regarding expected cluster specs

**Comment by ihoegen** [comment]
Looks like you still have issues with k8s.io/kops/tests/integration/create_cluster/shared_subnets_vpc_lookup/expected-v1alpha1.yaml and k8s.io/kops/tests/integration/create_cluster/shared_subnets_vpc_lookup/expected-v1alpha2.yaml

**Comment by ihoegen** [comment]
Looks like you need to run ./hack/update-bazel.sh

**Comment by integrii** [comment]
I ran it and it did nothing (running theme here), so I just manually modified the bezel file it talked about.  Hopefully that helps.  There also some complaint about the yaml stuff still.  I will look at it.

**Comment by ihoegen** [comment]
Looks like you still need work on
tests/integration/create_cluster/shared_subnets_vpc_lookup/expected-v1alpha1.yaml
tests/integration/create_cluster/shared_subnets_vpc_lookup/expected-v1alpha2.yaml
tests/integration/create_cluster/ha_gce/expected-v1alpha2.yaml
pkg/model/components/etcdmanager/tests/minimal/

File: pkg/apis/kops/cluster.go [hunk 1]
```diff
@@ -396,6 +397,10 @@ type EtcdClusterSpec struct {
 	Backups *EtcdBackupSpec `json:"backups,omitempty"`
 	// Manager describes the manager configuration
 	Manager *EtcdManagerSpec `json:"manager,omitempty"`
+	// MemoryRequest specifies the memory requests of each etcd container in the cluster.
+	MemoryRequest resource.Quantity `json:"memoryRequest,omitempty"`
+	// CPURequest specifies the cpu requests of each etcd container in the cluster.
+	CPURequest resource.Quantity `json:"cpuRequest,omitempty"`
```
Comment (KashifSaadat) [inline pkg/apis/kops/cluster.go]: Can you set these to `*resource.Quantity` rather than `resource.Quantity` (in the 3 cluster.go files) so to keep the same behaviour (spec) for existing clusters, unless explicitly set by the user. It will also pick out updates required to some tests as the nil values should be dropped.

Comment (integrii) [reply pkg/apis/kops/cluster.go]: In general, in my projects, I normally avoid the use of pointers because those nils always seem to come back to bite you, or require extra handling at some point.  This is not my project, though :-)

I tried to match the existing behavior by editing code in other places.  Check `protokube/pkg/protokube/etcd_cluster.go` line 194... `if c.MemoryRequest.IsZero() {` is used to determine if the user did not specify a quantity.  The existing defaults are used in that case.

Are you sure that the user's clusters will change if nothing was specified?  I specifically tried to write the code in such a way that users could ignore the option and functionality would not change.

Thank you for taking the time to look at this big PR!  It's a simple change, but many parts of the code are effected.

Comment (integrii) [reply pkg/apis/kops/cluster.go]: Understood!  It is a shame that the `resource.Quantity` does not marshal in such a way that `omitempty` will take effect.  I will work on switching these into pointers.

**Comment by integrii** [comment]
FYI this also fixes #6314 because i added `unset GOBIN` in the `hack/make-apimachinery.sh` script.

File: pkg/apis/kops/cluster.go (hunk 1 above)
Comment (integrii) [reply pkg/apis/kops/cluster.go]: @gambol99 this is done.

**Comment by KashifSaadat** [comment]
Hey @integrii, thanks for the updates!

Sorry to nitpick, just noticed that on a cluster spec update to include custom resource requests this is currently not picked up as a change when running `kops rolling-update cluster`. Meaning the instances would require a forced rolling-update.

To do so would require an update here: https://github.com/kubernetes/kops/blob/master/pkg/model/bootstrapscript.go#L159-L162

When the resource requests are included in the user-data for an instance, it will update the LC and ASG (in aws case), and mark the nodes as needing an update. To keep existing behaviour for people who haven't specified the new flags, you would need to perform a check here to see if the values are set, before including in the user-data output. This would then correctly mark the relevant nodes for a rolling-update only when the etcd spec is updated for the new options.

**Comment by integrii** [comment]
I took a shot at the change you mentioned @KashifSaadat - but I am not 100% I fully understand the code flow enough.  Can you look at my change to see if things look right?

**Comment by integrii** [comment]
I am having some issues with `make gofmt`.  I filed an issue for it: #6394

**Comment by integrii** [comment]
Hrmm... Should I just `retest`?  It says it was waiting for the cluster to come up.  That either means that my tests broke something, or the tests broke on their own. :-)

**Comment by integrii** [comment]
I am not sure what the best way to see into the failing test is here.  Something is going wrong which causes the cluster to not spin up in the `pull-kops-e2e-kubernetes-aws` test...

**Comment by KashifSaadat** [comment]
@integrii looks like some other unrelated failure in the kops e2e tests that are occurring

**Comment by integrii** [comment]
@KashifSaadat LGTM, more feedback, or squash? :-)  Tests work now!

**Comment by KashifSaadat** [comment]
That's great, thank you for addressing the comments @integrii and sorry about all the back and forth! Yeah if you could do a squash and we'll get this merged in 👍

**Comment by integrii** [comment]
Whew, okay that was a bit gory.  I had to back out a merge from master I did before I could rebase to squash commits because it wanted to rebase everything that had happened to master...

We should be down to just one commit from me and two from Ian, to maintain his contributions. 

I am now going to re-merge master and re-resolve conflicts since the rebase is complete.

**Comment by integrii** [comment]
@KashifSaadat This is squashed down quite a bit - good to go?

**Comment by integrii** [comment]
Finally I was able to squash all my contiguous commits between upstream merges.  Whoo!

**Comment by KashifSaadat** [comment]
That looked painful! Thanks for sorting out all the conflicts and squashing the commits :) I've dropped you a message on Slack in regards to the test fix, specifically in the `pkg/model/bootstrapscript.go` file.

**Comment by integrii** [comment]
@KashifSaadat you're my hero.  I made the change - lets see if builds pass!

**Comment by KashifSaadat** [comment]
All green! Thanks @integrii :+1: 

/lgtm
/approve

**Comment by integrii** [comment]
Excellent!!  I will try to make time to hit up the Calico specs next.
